packages/block-editor/src/store/reducer.js

```diff
@@ -922,161 +922,177 @@ export function isCaretWithinFormattedText( state = false, action ) {
 		}
-		case 'TOGGLE_SELECTION':
+	}
+
+	return state;
+}
+
+/**
+ * Reducer returning the block selection's start.
+ *
+ * @param {Object} state  Current state.
+ * @param {Object} action Dispatched action.
+ *
+ * @return {Object} Updated state.
+ */
+export function selectionStart( state = {}, action ) {
+	switch ( action.type ) {
+		case 'SELECTION_CHANGE':
 			return {
-				...state,
-				isEnabled: action.isSelectionEnabled,
+				clientId: action.clientId,
+				attributeKey: action.attributeKey,
+				offset: action.startOffset,
 			};
+		case 'RESET_SELECTION':
+			return action.selectionStart;
+		case 'MULTI_SELECT':
+			return { clientId: action.start };
+	}
+
+	return selection( state, action );
+}
+
+/**
+ * Reducer returning the block selection's end.
+ *
+ * @param {Object} state  Current state.
+ * @param {Object} action Dispatched action.
+ *
+ * @return {Object} Updated state.
+ */
+export function selectionEnd( state = {}, action ) {
+	switch ( action.type ) {
 		case 'SELECTION_CHANGE':
 			return {
-				...state,
-				start: {
-					clientId: action.clientId,
-					attributeKey: action.attributeKey,
-					offset: action.startOffset,
-				},
-				end: {
-					clientId: action.clientId,
-					attributeKey: action.attributeKey,
-					offset: action.endOffset,
-				},
+				clientId: action.clientId,
+				attributeKey: action.attributeKey,
+				offset: action.endOffset,
 			};
+		case 'RESET_SELECTION':
+			return action.selectionEnd;
+		case 'MULTI_SELECT':
+			return { clientId: action.end };
+	}
+
+	return selection( state, action );
+}
+
+/**
+ * Reducer returning whether the user is multi-selecting.
+ *
+ * @param {boolean} state  Current state.
+ * @param {Object}  action Dispatched action.
+ *
+ * @return {boolean} Updated state.
+ */
+export function isMultiSelecting( state = false, action ) {
+	switch ( action.type ) {
+		case 'START_MULTI_SELECT':
+			return true;
+
+		case 'STOP_MULTI_SELECT':
+			return false;
+	}
+
+	return state;
+}
+
+/**
+ * Reducer returning whether selection is enabled.
+ *
+ * @param {boolean} state  Current state.
+ * @param {Object}  action Dispatched action.
+ *
+ * @return {boolean} Updated state.
+ */
+export function isSelectionEnabled( state = true, action ) {
+	switch ( action.type ) {
+		case 'TOGGLE_SELECTION':
+			return action.isSelectionEnabled;
 	}
 
 	return state;
 }
 
+/**
+ * Reducer returning the intial block selection.
+ *
+ * Currently this in only used to restore the selection after block deletion.
+ * This reducer should eventually be removed in favour of setting selection
+ * directly.
+ *
+ * @param {boolean} state  Current state.
+ * @param {Object}  action Dispatched action.
+ *
+ * @return {?number} Initial position: -1 or undefined.
+ */
+export function initialPosition( state, action ) {
+	if ( action.type === 'SELECT_BLOCK' ) {
+		return action.initialPosition;
+	} else if ( action.type === 'REMOVE_BLOCKS' ) {
+		return state;
+	}
+}
```

Comment:
This reducer is a little unconventional, in that its default behaviour isn't to return `state`, but rather to implicitly return `undefined`. Is this the best way to proceed? Seems like it could increase the likelihood of mistakes in the future.

Reply:
How so? The reducer should store the initial position temporarily, until another action occurs. The only exception is `REMOVE_BLOCKS`. As mentioned above in the doc, the reducer should eventually be removed in favour of setting selection directly.

Reply:
This is more or less the same observation as in https://github.com/WordPress/gutenberg/pull/17482#discussion_r328165271

I'll defer to your judgment, as I don't want to block this.

Reply:
Ok, I'll add a comment.